### DISCUSSION TITLE
How to use git with multiple remote repositories?

### DISCUSSION THREAD

**Discussion by JohnBartlesAccount** [community/community#23289] [New to GitHub]
<p>   Hi I am trying to understand “<em>git remote add</em>”, and I am trying to understand it in the context of the below example marked “<em>Getting set up</em>”. I took the below example from the blog labeled, “<a href="https://mmikowski.github.io/git-cross-origin/" rel="nofollow noopener">Using git with multiple remote repositories</a>”.</p>
<p>From the line below which reads “git checkout -b ltullman-master”, what branch is the newly created  branch "<em>ltullman-master" based</em> on <strong>?</strong></p>
<p>From the line below which reads “git fetch ltullman master”, how can the current checkout branch “ltullman-master” fetch anything at all from the address “<a href="mailto:git@github.com">git@github.com</a>:ltullman/taffydb.git” since “itullman-master” is originally based on the clone from the address “<a href="mailto:git@github.com">git@github.com</a>:typicaljoe/taffydb.git”<strong>?</strong></p>
<p>…Wouldn’t there be a huge number of conflicts<strong>?</strong></p>
<p>  In other words, it seems to me that the current checkout branch is fetching from one address but it’s originally based on and cloned from another address, and that seems like it shouldn’t work.</p>
<p>=======Getting set up================</p>
<p>Below I illustrate checking out taffydb on the master branch, and then set up two alternate branches using two different source repositories.</p>
<p>First, let’s get the canonical taffydb repository:</p>
<pre><code>git clone git@github.com:typicaljoe/taffydb.git taffydb
</code></pre>
<p>Now let’s enter that directory and have a look around:</p>
<pre><code>cd taffydb
git remote -v
  # origin git@github.com:typicaljoe/taffydb.git (fetch)
  # origin git@github.com:typicaljoe/taffydb.git (push)
</code></pre>
<p>Let’s add our upstream branch to ltullman.</p>
<pre><code># create new branch
git checkout -b ltullman-master

# add remote repository
git remote add ltullman git@github.com:ltullman/taffydb.git

# fetch the master branch contents
git fetch ltullman master

# and now set upstream
git branch --set-upstream-to=ltullman/master
</code></pre>

**Comment by Wabri** [comment, marked as answer]
<p>Hi <a class="mention" href="/johnbartlesaccount">@johnbartlesaccount</a> ,</p>
<p> Welcome to the GitHub Community forum!</p>
<blockquote>
<p>From the line below which reads “git checkout -b ltullman-master”, what branch is the newly created  branch "<em>ltullman-master" based</em> on <strong>?</strong></p>
</blockquote>
<p>It is based on the commit where you are. Remember that git is a decentralized version controll sysyem so the repository in your local machine is different from the repository on remote.</p>
<p>For example you clone a repository from remote where there are already some commits:</p>
<p>                   A—B---C—D     remote/master</p>
<p>                             II       (clone)</p>
<p>                   A—B---C—D     local/master</p>
<p>You are now in the HEAD (commit D) of the master branch in your local repository. Now when you create a new branch it’s append something like this:</p>
<p>                                           D  new.branch</p>
<p>                                         /</p>
<p>                   A—B---C—D   master</p>
<p>But this is your local repository so the branch is not based in any of the remote repository. In your case <strong>new.branch</strong> is <strong>itullman-master.</strong></p>
<p>To merge this branch in one of the repository you need to do:</p>
<pre><code>git push &lt;remote&gt; &lt;new.branch&gt;
</code></pre>
<p>Where remote is the remote where you want this branch.</p>
<blockquote>
<p>From the line below which reads “git fetch ltullman master”, how can the current checkout branch “ltullman-master” fetch anything at all from the address “<a href="mailto:git@github.com">git@github.com</a>:ltullman/taffydb.git” since “itullman-master” is originally based on the clone from the address “<a href="mailto:git@github.com">git@github.com</a>:typicaljoe/taffydb.git”<strong>?</strong></p>
</blockquote>
<p>If I understood the question, the answer is: if there is a branch named <strong>master</strong> on Itullman it fetchs that on the new branch <strong>itullman-master</strong>.</p>
<blockquote>
<p>…Wouldn’t there be a huge number of conflicts<strong>?</strong></p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>  In other words, it seems to me that the current checkout branch is fetching from one address but it’s originally based on and cloned from another address, and that seems like it shouldn’t work.</p>
</blockquote>
<p>You need to do the fetch to prevent the conflict in the master based on <strong>remote origin</strong>. In fact you create a the branch <strong>ltullman-master</strong> where now you try to do the fetch of the master branch of <strong>remote Itullman,</strong> this can generate a huge conflict or not. In the first case you “save” the master of your local repository because the fetch is made into the <strong>itullman-master.</strong></p>
<p>Take the example above, now you have this:</p>
<p>                   A—B---C—D     origin/master</p>
<p>                             II       (clone)</p>
<p>                   A—B---C—D     local/master</p>
<p>                                        &lt;/font&gt;</p>
<p>                                         D   local/itullman-master</p>
<p>                   X—X---X—X     itullman/master</p>
<p>You are in the <strong>HEAD</strong> of <strong>local</strong> repository in <strong>itullman-master</strong> branch and you do a fetch, this is what happen:</p>
<ol>
<li>if the itullman remote have the same history</li>
</ol>
<p>                   A—B---C—D     local/itullman-master</p>
<p>                             II       (fetch)</p>
<p>                   A—B---C—D     itullman/master</p>
<p><strong>Fetch result = Already up to date</strong></p>
<p>The history are the same.</p>
<ol start="2">
<li>if the itullman remote have the same history</li>
</ol>
<p>                   A—B---C—D     local/itullman-master</p>
<p>                             II       (fetch)</p>
<p>                   Z—Y---X—W     itullman/master</p>
<p><strong>Fetch result = Conflict</strong></p>
<p>This because the history of master of itullman remote repository are different from the itullman-master of local repository.</p>
<ol start="3">
<li>if the itullman remote have the same history</li>
</ol>
<p>                   A—B---C—D     local/itullman-master</p>
<p>                             II       (fetch)</p>
<p>                   A—B---C—D---E—F     itullman/master</p>
<p><strong>Fetch result = local/itullman-master are behind of 2 commits</strong></p>
<p>(Similar if local/itullman-master are ahead)</p>
<p>If you haven’t understood something, ask for more.</p>
<p>Hope this helps :slightly_smiling_face:  </p>
<p>-Gabriele-</p>
<blockquote>
<p><strong>Mark helpful posts with Accept as Solution to help other users locate important info. Don’t forget to give Kudos for great content!</strong></p>
</blockquote>

**Comment by JohnBartlesAccount** [comment]
<p>Thanks  <a class="mention" href="/wabri">@wabri</a>  for the answer. You seemed to resolve my confusion. Your answer seems to imply that if a local branch has multiple upstream addresses then the branches on the upstream addresses, from which your local branch fetches, SHOULD have a fairly similar history OTHERWISE multiple upstream addresses(i.e. remotes) should not be used.</p>
<p> I was just introduced to the idea of multiple upstream addresses a few days ago. From my limited perspective, ANYTHING could have been possible and anything could have been a reason for having multiple upstream addresses, and there could have been ANY number of right or wrong ways of managing and handling multiple upstream addresses.  Looking back, however, I guess it was obvious that in order to use multiple remotes for a single local branch those remote branches must be, at least, fairly similar.</p>
<p>Ok, buddy, thanks for the resolve.</p>

**Comment by twildeboer** [comment]
<p>The commands from the example didn’t quite work the way I wanted. Using the same repo names from the example, this is what did work (and makes more sense to me):</p>
<pre><code># Clone
git clone git@github.com:typicaljoe/taffydb.git taffydb

# Enter the repo
cd taffydb

# (No need to create the new branch yet)

# Add the new remote
git remote add ltullman git@github.com:ltullman/taffydb.git

# Fetch branches from the new remote *without specifying a branch*
git fetch ltullman

# Create a new branch (ltullman-master) from the new remote's master
# and automatically setup the tracking (assuming you're configured to do so)
git checkout -b ltullman-master ltullman/master
</code></pre>
